test/TestCaseReader.cpp

```diff
@@ -136,6 +144,54 @@ pair<SourceMap, size_t> TestCaseReader::parseSourcesAndSettingsWithLineNumber(is
 				if (sources.count(currentSourceName))
 					BOOST_THROW_EXCEPTION(runtime_error("Multiple definitions of test source \"" + currentSourceName + "\"."));
 			}
+			else if (boost::algorithm::starts_with(line, externalSourceDelimiterStart) && boost::algorithm::ends_with(line, sourceDelimiterEnd))
+			{
+				string externalSourcePath = boost::trim_copy(line.substr(
+					externalSourceDelimiterStart.size(),
+					line.size() - sourceDelimiterEnd.size() - externalSourceDelimiterStart.size()
+				));
+				string externalSourceName = externalSourcePath;
+
+				// Does the external source define a remapping?
+				size_t remappingPos = externalSourcePath.find('=');
+				if (remappingPos != string::npos)
+				{
+					externalSourceName = boost::trim_copy(externalSourcePath.substr(0, remappingPos));
+					externalSourcePath = boost::trim_copy(externalSourcePath.substr(remappingPos + 1));
+				}
+
+				string externalSourceContent;
+				fs::path parent_path(m_filename);
+				parent_path = parent_path.normalize().remove_filename();
+
+				fs::path path = parent_path / fs::path(externalSourcePath);
+
+				if (fs::exists(path))
+					externalSourceContent = util::readFileAsString(path.c_str());
+				else
+					BOOST_THROW_EXCEPTION(runtime_error(string("External Source '"+ externalSourcePath +"' not found.")));
+
+				if (!externalSourceName.empty())
+				{
+					ErrorList errorList;
+					ErrorReporter errorReporter(errorList);
+					Parser parser(errorReporter, langutil::EVMVersion::berlin());
+					shared_ptr<Scanner> scanner = make_shared<Scanner>(CharStream(externalSourceContent, externalSourceName));
+					ASTPointer<SourceUnit> sourceUnit = parser.parse(scanner);
+					if (sourceUnit)
+						for (auto const& node: sourceUnit->nodes())
+							if (auto import = dynamic_cast<ImportDirective const*>(node.get()))
+							{
+								fs::path parent(externalSourcePath);
+								parent = parent.parent_path();
+								string importedSourceContent = util::readFileAsString(parent_path.string() + "/" + parent.string() + "/" + import->path());
+								sources[parent.string() + "/" + import->path()] = importedSourceContent;
```

Comment:
I think it is enough to be able just to point to one solidity file that imports the needed sources. We also could do this recursively, but I think it sufficient to just point to a file that imports all needed external sources.

Reply:
Maybe `(parent / import->path()).string()`?

Reply:
I had that before exactly like this. The problem is that solidity only supports the unix style path separators. If we use  the / operator here, we will have backslashes for windows, that we don't want in this case.

Reply:
> You mean that if, say `a.sol` imports `b.sol` and `b.sol` imports `c.sol`, it's enough to put `a.sol` and `b.sol` in `sources` and the `CompilerStack` will just find the sources on disk? If so, I think this might not work. Path resolution in `CompilerStack` is quite different than here.
> 

I will add some tests to verify the behaviour. But yes, I thought it could work like this. Let me check.

> If you do `(parent / import->path()).generic_string()`, you will get UNIX backslashes. This is actually how `CompilerStack` does it too.

Ah nice!

Reply:
> I will add some tests to verify the behaviour. But yes, I thought it could work like this. Let me check.

I mean, it would work in the simplest case where you have the files in a single directory, importing each other and the working directory is that test directory. It will break when you start using relative imports or subdirectories. And the working directory is usually the repo root so that will also interfere.

Reply:
Actually, I have an idea. You could create a custom file loader (the default is [FileReader::readFile()](https://github.com/ethereum/solidity/blob/develop/libsolidity/interface/FileReader.cpp#L45-L94)) that looks for files in the test directory instead of in the current working directory. The loader is basically just lambda and you could ignore all the complications of base path.

You could just reuse `FileReader` and only set a custom base path which would be even simpler and you would avoid duplicating the logic.

Reply:
I'm wondering whether this is too complicated. I think complex import situations could be solved by defining a remapping.

Reply:
Well, reusing `FileReader` class for this would just require creating a new instance and passing it to `CompilerStack` when it's created. So not that complicated. I think that designating the test directory as the working directory actually does make a lot of sense. It would probably make nested imports "just work" as you originally intended.

But well, we could also just ignore these nested imports in this PR and deal with them separately.

Reply:
Actually, forget about reusing `FileReader`. I realized that supporting relative imports and basically any kind of imports is easy as long as we just stop at the virtual filesystem level and never go to the actual filesystem. So please disregard my suggestion above and just look at:
- https://github.com/ethereum/solidity/pull/11110#discussion_r617047235
- https://github.com/ethereum/solidity/pull/11110#issuecomment-823648499